rust/src/collaboration/databasesync.rs

```diff
@@ -0,0 +1,708 @@
+        )
+    };
+    ptr::NonNull::new(result)
+        .map(|raw| unsafe { RemoteFile::from_raw(raw) })
+        .ok_or(())
+}
+
+/// Test if a database is valid for use in collaboration
+pub fn is_collaboration_database(database: &Database) -> bool {
+    unsafe { BNCollaborationIsCollaborationDatabase(database.as_raw()) }
+}
+
+/// Get the Remote for a Database
+pub fn get_remote_for_local_database(database: &Database) -> Result<Option<Remote>, ()> {
+    let mut value = ptr::null_mut();
+    let success =
+        unsafe { BNCollaborationGetRemoteForLocalDatabase(database.as_raw(), &mut value) };
+    success
+        .then(|| ptr::NonNull::new(value).map(|handle| unsafe { Remote::from_raw(handle) }))
+        .ok_or(())
+}
+
+/// Get the Remote for a BinaryView
+pub fn get_remote_for_binary_view(bv: &BinaryView) -> Result<Option<Remote>, ()> {
+    let Some(db) = bv.file().database() else {
+        return Ok(None);
+    };
+    get_remote_for_local_database(&db)
+}
+
+/// Get the Remote Project for a Database, returning the Remote project from one of the
+/// connected remotes, or None if not found or if projects are not pulled
+pub fn get_remote_project_for_local_database(
+    database: &Database,
+) -> Result<Option<RemoteProject>, ()> {
+    let mut value = ptr::null_mut();
+    let success =
+        unsafe { BNCollaborationGetRemoteProjectForLocalDatabase(database.as_raw(), &mut value) };
+    success
+        .then(|| ptr::NonNull::new(value).map(|handle| unsafe { RemoteProject::from_raw(handle) }))
+        .ok_or(())
+}
+
+/// Get the Remote File for a Database
+pub fn get_remote_file_for_local_database(database: &Database) -> Result<Option<RemoteFile>, ()> {
+    let mut value = ptr::null_mut();
+    let success =
+        unsafe { BNCollaborationGetRemoteFileForLocalDatabase(database.as_raw(), &mut value) };
+    success
+        .then(|| ptr::NonNull::new(value).map(|handle| unsafe { RemoteFile::from_raw(handle) }))
+        .ok_or(())
+}
+
+/// Add a snapshot to the id map in a database
+pub fn assign_snapshot_map(
+    local_snapshot: &Snapshot,
+    remote_snapshot: &CollabSnapshot,
+) -> Result<(), ()> {
+    let success = unsafe {
+        BNCollaborationAssignSnapshotMap(local_snapshot.as_raw(), remote_snapshot.as_raw())
+    };
+    success.then_some(()).ok_or(())
+}
+
+/// Get the remote snapshot associated with a local snapshot (if it exists)
+pub fn get_remote_snapshot_from_local(snap: &Snapshot) -> Result<Option<CollabSnapshot>, ()> {
+    let mut value = ptr::null_mut();
+    let success = unsafe { BNCollaborationGetRemoteSnapshotFromLocal(snap.as_raw(), &mut value) };
+    success
+        .then(|| ptr::NonNull::new(value).map(|handle| unsafe { CollabSnapshot::from_raw(handle) }))
+        .ok_or(())
+}
+
+/// Get the local snapshot associated with a remote snapshot (if it exists)
+pub fn get_local_snapshot_for_remote(
+    snapshot: &CollabSnapshot,
+    database: &Database,
+) -> Result<Option<Snapshot>, ()> {
```

Comment:
`Result<Option<Snapshot>, ()>` -> `Option< Snapshot >`

Reply:
This need to be `Result<Option<T>, ()>` or `Option<Option<T>>`. Similar to above, need to differentiate an error from a value not found.